### PR TITLE
Pylibcudf: pack and unpack

### PR DESCRIPTION
**Pull request by madsbk**
Adding python bindings to [`cudf::pack()`](https://docs.rapids.ai/api/libcudf/legacy/group__copy__split#ga86716e7ec841541deb6edc7e91fcb9e4), [`cudf::unpack()`](https://docs.rapids.ai/api/libcudf/legacy/group__copy__split#ga1d62a18c2e6f087a92289c63693762cc), and [`cudf::packed_columns`](https://docs.rapids.ai/api/libcudf/legacy/structcudf_1_1packed__columns).

This is the first step to support serialization of cudf.polars' IR.

cc. @wence- @rjzamora 

### Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by madsbk** [comment]
Sorry @Matt711, didn't see https://github.com/rapidsai/cudf/pull/16953. 

> xref #16953. Do you also plan to plumb your changes through cudf python (eg. [copying.pyx](https://github.com/rapidsai/cudf/blob/09ed2105b841fe29be75af8b0d5a41fc09e7b6ac/python/cudf/cudf/_lib/copying.pyx#L486))

No, not in this PR. The plan is to include the bindings needed to serialize cudf.polars' [`DataFrame`](https://github.com/rapidsai/cudf/blob/09ed2105b841fe29be75af8b0d5a41fc09e7b6ac/python/cudf_polars/cudf_polars/containers/dataframe.py#L31).

**Comment by Matt711** [comment]
> Sorry @Matt711, didn't see #16953.
> 
> > xref #16953. Do you also plan to plumb your changes through cudf python (eg. [copying.pyx](https://github.com/rapidsai/cudf/blob/09ed2105b841fe29be75af8b0d5a41fc09e7b6ac/python/cudf/cudf/_lib/copying.pyx#L486))
> 
> No, not in this PR. The plan is to include the bindings needed to serialize cudf.polars' [`DataFrame`](https://github.com/rapidsai/cudf/blob/09ed2105b841fe29be75af8b0d5a41fc09e7b6ac/python/cudf_polars/cudf_polars/containers/dataframe.py#L31).

No worries, I'll probably hold off on #16953, and get any missing pieces from `contiguous_split.hpp` in there after your PR gets merged.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (madsbk) [inline python/pylibcudf/pylibcudf/contiguous_split.pyx:79]: Here we copy the `table_view` return from `cpp_unpack`. Is it possible to avoid this copy?
The problem is that [`Table.from_table_view()`](https://github.com/rapidsai/cudf/blob/219ec0e7fbff37d7387d25e93510b55a8782e2bf/python/pylibcudf/pylibcudf/table.pyx#L63C38-L63C48) takes `Table owner`, which we don't have.


cc. @vyasr

Comment (wence-) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx:79]: We can extend that to take arbitary owning objects I suspect, if we can figure out a way to let `from_column_view` similarly take arbitrary owning objects.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (bdice) [inline python/pylibcudf/pylibcudf/contiguous_split.pyx:154]:
```suggestion
            raise ValueError("Expected an empty gpu_data from unpacking an empty table")
```

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (wence-) [inline python/pylibcudf/pylibcudf/contiguous_split.pyx:127]: If we don't think we can solve this copy here, can we open an issue discussing _how_ to solve it?

Comment (madsbk) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx:129]: done: https://github.com/rapidsai/cudf/issues/17040

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (mroeschke) [inline python/pylibcudf/pylibcudf/contiguous_split.pyx:21]: Just noting that we'll need to explicitly include numpy as a required dependency for pylibcudf.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (wence-) [inline python/pylibcudf/pylibcudf/contiguous_split.pyx:66]: To avoid the numpy dependence, could `PackedColumns` just advertise the buffer protocol directly? Then you could say `memoryview(packed_columns)` and get the metadata. Or return an object that advertises the buffer protcol: https://cython.readthedocs.io/en/latest/src/userguide/buffer.html

In python 3.12 and later we could just do that via the python `__buffer__` attribute, but we can easily do it in cython.

Comment (madsbk) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx]: Implemented `HostBuffer` that owns the data and implement the buffer protocol.

And we now have `PackedColumns.release()` to get the buffers. This means that we can free the two buffers individually.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (vyasr) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx:192]: Continuing discussion in https://github.com/rapidsai/cudf/issues/17040.

**Review by mroeschke — APPROVED** [review]

File: python/pylibcudf/pylibcudf/tests/test_contiguous_split.py
Comment (wence-) [inline python/pylibcudf/pylibcudf/tests/test_contiguous_split.py:13]:
```suggestion
    pa.table({"a": ["a", "bb", "ccc"]}),
    pa.table({"a": [1, 2, None], "b": [None, 3, 4]})
    pa.table({"a": [["a", "b"], ["cde"]], "b": {"alpha": [1, 2], "beta": None}})
```

Can we test with nullable columns and nested types as well?

File: python/pylibcudf/pylibcudf/tests/test_contiguous_split.py
Comment (wence-) [inline python/pylibcudf/pylibcudf/tests/test_contiguous_split.py:23]: This one will fail too, because it also has a struct column.

**Review by wence- — APPROVED** [review]
One more test param wants xfailed, I _think_ (let's wait for it to run through)

File: python/pylibcudf/pylibcudf/contiguous_split.pyx
Comment (wence-) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx]: :+1: thanks!

File: python/pylibcudf/pylibcudf/tests/test_contiguous_split.py
Comment (wence-) [reply python/pylibcudf/pylibcudf/tests/test_contiguous_split.py:23]: Ah, no, because `assert_frame_eq` sets up the metadata correctly.

**Pull request merged**
